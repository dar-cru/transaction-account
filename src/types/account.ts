export enum AccountStatus {
  ACTIVE = "active",
  PENDING = "pending"
}

export type Account = {
  type: string;
  id: string;
  name: string;
  status: AccountStatus;
  balance: number;
};