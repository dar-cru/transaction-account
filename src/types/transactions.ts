export enum TransactionType {
  DEBIT = "debit",
  CREDIT = "credit",
}

export type Transaction = {
  name: string;
  amount: number;
  type: TransactionType;
  dateTime: string;
};

export type TransactionResponse = {
  accountNumber: number;
  id: number;
  transactions: Transaction[];
};