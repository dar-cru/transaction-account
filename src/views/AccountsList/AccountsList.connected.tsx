import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../redux/rootReducer";
import { fetchAccounts } from "../../slices/accountsSlice";
import AccountsList from "./AccountsList";

export const Connected = () => {
  const { accounts, isLoading, error } = useSelector(
    (state: RootState) => state.account
  );
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchAccounts());
  }, [dispatch]);

  return <AccountsList accounts={accounts} isLoading={isLoading} errorMsg={error} />;
};

export default Connected;
