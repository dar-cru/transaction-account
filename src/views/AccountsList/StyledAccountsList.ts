import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  max-width: 800px;
  flex-flow: wrap;
  margin: 0 auto;
`;

export const CardContainer = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;

  @media screen and (min-width: 992px) {
    width: 50%;
  }
`;
