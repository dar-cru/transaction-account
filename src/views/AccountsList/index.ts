import Connected from './AccountsList.connected';

export default Connected;