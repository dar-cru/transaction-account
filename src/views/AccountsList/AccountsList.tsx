import { MouseEvent } from "react";
import { Account } from "../../types";
import Card from "../../components/Card/Card";
import { useHistory } from "react-router-dom";
import InPageAlert from "../../components/InPageAlert";
import { Container, CardContainer } from "./StyledAccountsList";

export type AccountsProps = {
  accounts: Account[];
  isLoading?: boolean;
  errorMsg: string | null;
};

const AccountsList = ({ accounts, isLoading, errorMsg }: AccountsProps) => {
  const history = useHistory();
  const viewTransactions = (event: MouseEvent<HTMLButtonElement>) => {
    const accountNumber = event.currentTarget.value;
    history.push({
      pathname: "/transactions",
      state: { accountNumber: accountNumber },
    });
  };

  return (
    <Container>
      {isLoading && "Loading..."}
      {accounts &&
        !errorMsg &&
        !isLoading &&
        accounts.map((account) => {
          return (
            <CardContainer key={`card-for-account-${account.name}`}>
              <Card {...account} onClick={viewTransactions} />
            </CardContainer>
          );
        })}
      {accounts?.length < 1 && !isLoading && (
        <InPageAlert
          type="error"
          message="There was an error retrieving your accounts"
        />
      )}
    </Container>
  );
};

export default AccountsList;
