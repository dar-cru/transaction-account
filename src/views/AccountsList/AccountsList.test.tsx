import React from "react";
import { render, RenderResult } from "@testing-library/react";
import AccountsList from "./AccountsList";
import { Account, AccountStatus } from "../../types";

const accounts: Account[] = [
  {
    name: "Progress Saver",
    type: "savings",
    id: "1",
    status: AccountStatus.ACTIVE,
    balance: 280500.22,
  },
  {
    name: "Every day account",
    type: "debit",
    id: "2",
    status: AccountStatus.PENDING,
    balance: 0,
  },
  {
    name: "Home Loan",
    type: "mortgage",
    id: "3",
    status: AccountStatus.PENDING,
    balance: -600000.22,
  },
  {
    name: "Qantas Black",
    type: "credit",
    id: "4",
    status: AccountStatus.PENDING,
    balance: -340.22,
  },
];

describe("<AccountsList />", () => {
  test("renders the login form if there is no authentication", () => {
    const tree: RenderResult = render(<AccountsList accounts={accounts} isLoading={false} errorMsg={null} />);

    const { container } = tree;
    expect(container).toMatchSnapshot();
  });
});
