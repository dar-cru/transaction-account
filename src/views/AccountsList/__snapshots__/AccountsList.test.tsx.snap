// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountsList /> renders the login form if there is no authentication 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 10px;
  margin: 20px;
  cursor: pointer;
  border: none;
  border-radius: 5px;
  box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
  -webkit-transition: 0.3s;
  transition: 0.3s;
  min-width: 320px;
  min-height: 104px;
}

.c0:hover {
  box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
}

.c4 {
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  color: crimson;
  margin-bottom: -10px;
}

.c1 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  margin: 10px 0;
}

.c2 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  margin: 0 0 6px;
}

.c3 {
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  color: dodgerblue;
}

<div>
  <div
    style="display: flex; width: 800px; flex-flow: wrap; margin: 0px auto;"
  >
    <div
      style="width: 50%;"
    >
      <button
        class="c0"
        value="1"
      >
        <h3
          class="c1"
        >
          Progress Saver
        </h3>
        <h4
          class="c2"
        >
          savings
           account
        </h4>
        <span>
          $280,500.22
        </span>
        <span
          class="c3"
        >
          View Transactions 
          &gt;
        </span>
      </button>
    </div>
    <div
      style="width: 50%;"
    >
      <button
        class="c0"
        value="2"
      >
        <span
          class="c4"
        >
          <i>
            Pending
          </i>
        </span>
        <h3
          class="c1"
        >
          Every day account
        </h3>
        <h4
          class="c2"
        >
          debit
           account
        </h4>
        <span>
          $0.00
        </span>
      </button>
    </div>
    <div
      style="width: 50%;"
    >
      <button
        class="c0"
        value="3"
      >
        <span
          class="c4"
        >
          <i>
            Pending
          </i>
        </span>
        <h3
          class="c1"
        >
          Home Loan
        </h3>
        <h4
          class="c2"
        >
          mortgage
           account
        </h4>
        <span>
          -$600,000.22
        </span>
      </button>
    </div>
    <div
      style="width: 50%;"
    >
      <button
        class="c0"
        value="4"
      >
        <span
          class="c4"
        >
          <i>
            Pending
          </i>
        </span>
        <h3
          class="c1"
        >
          Qantas Black
        </h3>
        <h4
          class="c2"
        >
          credit
           account
        </h4>
        <span>
          -$340.22
        </span>
      </button>
    </div>
  </div>
</div>
`;
